### ISSUE TITLE
ST_IsValidReason() throws exception

### ISSUE THREAD

**Issue by edigonzales** [orbisgis/h2gis#1194]
Polygon contains a ring self-intersection. `ST_IsValidReason()` throws an exception. h2 db is attached.

dbeaver output:
```
SQL Error [90105] [90105]: Exception calling user-defined function: "isValidReason(POLYGON ((2605983.883 1258545.124, 2605984.428 1258547.158, ... , 2605260.802 1258848.107))): 'org.locationtech.jts.geom.LineString org.locationtech.jts.geom.Polygon.getExteriorRing()'"; SQL statement:
SELECT 
	st_isvalidreason(geometrie)
FROM 
	NUTZUNGSPLANUNG_GRUNDNUTZUNG 
WHERE 
	st_isvalid(geometrie) = FALSE
 [90105-197]
  org.h2.jdbc.JdbcSQLException: Exception calling user-defined function: "isValidReason(POLYGON ((2605983.883 1258545.124, 2605984.428 1258547.158, ..., 2605260.802 1258848.107))): 'org.locationtech.jts.geom.LineString org.locationtech.jts.geom.Polygon.getExteriorRing()'"; SQL statement:
SELECT 
	st_isvalidreason(geometrie)
FROM 
	NUTZUNGSPLANUNG_GRUNDNUTZUNG 
WHERE 
	st_isvalid(geometrie) = FALSE
 [90105-197]
  Exception calling user-defined function: "isValidReason(POLYGON ((2605983.883 1258545.124, 2605984.428 1258547.158, ..., 2605260.802 1258848.107))): 'org.locationtech.jts.geom.LineString org.locationtech.jts.geom.Polygon.getExteriorRing()'"; SQL statement:
SELECT 
	st_isvalidreason(geometrie)
FROM 
	NUTZUNGSPLANUNG_GRUNDNUTZUNG 
WHERE 
	st_isvalid(geometrie) = FALSE
 [90105-197]
    'org.locationtech.jts.geom.LineString org.locationtech.jts.geom.Polygon.getExteriorRing()'
    'org.locationtech.jts.geom.LineString org.locationtech.jts.geom.Polygon.getExteriorRing()'
```

[2476.mv.db.zip](https://github.com/orbisgis/h2gis/files/6057364/2476.mv.db.zip)

**Comment by edigonzales** [comment]
It seems to be an issue with the `Polygon.getExteriorRing()` method.  I get a similar error with

```
SELECT 
	ST_ExteriorRing(geometrie),
	*
FROM 
	NUTZUNGSPLANUNG_GRUNDNUTZUNG 
WHERE 
	t_ili_tid = '57a4f968-4e9b-11e8-b58f-6c3be52716b4'
```

Is this a corner case or does the jts method not work with self ring intersections. In latter case it cannot be used for `st_isvalidreason`, `st_makevalid', ...

**Comment by ebocher** [comment]
Hi,

Can you please share the whole geometry in WKT ?

**Comment by edigonzales** [comment]
Geometry as wkt: [polygon.wkt.zip](https://github.com/orbisgis/h2gis/files/6060384/polygon.wkt.zip)

Strange but on my office PC `ST_IsValidReason`, `ST_MakeValid` works. `ST_ExteriorRing` throws a different error:

```
Error occurred during SQL query execution

Reason:
SQL Error [50000] [HY000]: General error: "java.lang.IllegalArgumentException: Invalid ordinate index: 2"; SQL statement:
SELECT 
    --ST_MakeValid(geometrie),
    --ST_IsValidReason(geometrie),
    ST_ExteriorRing(geometrie),
    --ST_AsWkt(geometrie),
    *
FROM 
    NUTZUNGSPLANUNG_GRUNDNUTZUNG 
WHERE 
    t_ili_tid = '57a4f968-4e9b-11e8-b58f-6c3be52716b4'
 [50000-197]
```

**Comment by ebocher** [comment]
Thanks.
JTS.1.16 return an error at this position  POINT (2605853.029  1258771.176).
I will check H2GIS function

**Comment by edigonzales** [comment]
I checked the same query back home and I still get the original error message which is different from the one on my office pc.  Can I check e.g. the jar versions with some sql quey? dbeaver tells me twice H2 1.4.197.

**Comment by ebocher** [comment]
If you are using H2GIS with dbeaver there is two drivers :  one that corresponds to the 1.5 and use H2 1.4.197 and second that uses the H2GIS beta 2.0 (master) with a compiled version of H2 (2.0.201-SNAPSHOT). H2GIS beta 2.0 is  quite different than 1.5 in terms of geometry storage, syntax to declare a geometry table (aligned with postgis), JTS dependency...
This may be the cause of your problem.

**Comment by ebocher** [comment]
Is this issue still open ?

**Comment by edigonzales** [comment]
No, does not seem so. 

I created the h2gis database again with https://github.com/claeis/ili2db/tree/master/ili2h2gis which uses h2gis 1.5. Then made a copy of the database and opened it in dbeaver. `ST_IsValidReason` finds the ring self intersection without any exceptions and/or query errors.

I wrote some java test application which connects with jdbc. And `ST_IsValidReason` works flawlessly.

Perhaps I messed it up earlier with opening the database with different driver versions.
